### ISSUE TITLE
Pull fetches top-level `@include`/`@skip` into condition nodes

### ISSUE THREAD

**Issue by pcmanus** [apollographql/federation#2251]
This ticket is essentially about porting the feature added in #1113 for federation 1 in federation 2, but adapting it to federation 2 specific.

Essentially, when `@include`/`@skip` are involved, it can happen that some `FetchNode` in the plan ends up with some top-level(s) `@include`/`@skip`, meaning that the whole fetch is conditional on some condition(s). In such case, we can pull up those conditions out of the `FetchNode` selection set and into a specific query plan construct so that the gateway/router can possibly entirely avoid the fetch (saving a bit of load on the underlying service in that case).

I'll note that federation 1 implementation of this uses a new somewhat adhoc `inclusionConditions` parameter to `FetchNode`, but in federation 2, we added generic `ConditionNode` with the implementation of `@defer`, so we should reuse this (we might need to generalize slightly was we have to handle this cleanly).

**Comment by pcmanus** [comment]
> so we should reuse this (we might need to generalize slightly was we have to handle this cleanly).

And to clarify what I mean by "generalize" above: currently the `ConditionNode.condition` field is a string but is only used to contain a single boolean variable name. It would be useful here to have it support a more general notion of boolean "expressions", stuffs like `condition: "$v1 && $v2 && !$v3)"`. Which means agreeing on the grammar of what is allowed there. Longer term, it would be nice to support a relatively general form of boolean expression, something like:
```
condition ::= term ( '||' term )*
term      ::= neg  ( '&&' neg  )*
neg       ::= '!'? factor
factor    ::= var | '(' condition ')'
var       := '$' <variable name>
```
but to support `@include` and `@skip` here, I think we'll only need AND conditions and negation on variables, so if we feel like supporting the generic case is too complex initially, we could restrict it to:
```
condition :: = term ('&&' term)*
term      ::= '!'? '$' <variable name>
```
which is certainly a bit simpler to parse (you can just split on `&&` and just have to look for `!` on the result).

**Comment by pcmanus** [comment]
Pushed #2424 for this. It's essentially ready but I have 1-2 things I want to double check before calling it ready (details in the PR).

> currently the ConditionNode.condition field is a string but is only used to contain a single boolean variable name. It would be useful here to have it support a more general notion of boolean "expressions"

Fwiw, I bailed out of that idea of mine in the PR after all. As mentioned at the end of the previous comment, for `@include` and `@skip`, all we need to handle is `AND` on some conditions, and that can easily be done by using a simple cascade of `ConditionNode`, and I ended up doing that as that keeps things simpler for now. I still wouldn't be surprised if a time comes where supporting more complex boolean expressions becomes genuinely needed, but let's wait until such time.
